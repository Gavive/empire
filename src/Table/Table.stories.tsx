import React from "react";

import { ThemeProvider } from "styled-components";
import { withKnobs, optionsKnob } from "@storybook/addon-knobs";
import { withThemes } from 'storybook-addon-themes/react';


import Table from "./Table";

const Decorator = (props) => {
  let theme = lightTheme;

  if (props.theme && props.theme.name === "dark") {
    theme = darkTheme;
  }

  return (
    <ThemeProvider theme={theme}>
      {props.children}
    </ThemeProvider>
  )
}

export default {
  component: Table ,
  title: "Table",
  decorators: [withKnobs, withThemes],
  parameters: {
    themes: {
      Decorator,
      list: [
        {
          name: "light",
          color: "#fff",
          default: true
        },
        {
          name: "dark",
          color: "#121212"
        }
      ]
    },
    controls: { expanded: true }
  }
};

type Property = {
  name: string;
  bedBath: string;
  boughtOn: string;
  cost: number;
  equity: number;
}

const darkTheme = {
  name: "dark",
  color: "white"
};

const lightTheme = {
  name: "light",
  color: "black"
};

export const Default = (args) => (
  <Table<Property>
    {...args}
  />
);

Default.args = {
  columns: [
    {
      key: "name",
      label: "üè° Property"
    },
    {
      key: "bedBath",
      label: "üöø Bed/Bath"
    },
    {
      key: "boughtOn",
      label: "üìÖ Bought on"
    },
    {
      key: "cost",
      label: "üí∞ Purchase price"
    },
    {
      key: "equity",
      label: "üëç Equity"
    }
  ],
  data: [
    {
      name: "402 S 43rd St",
      bedBath: "3 Bed, 1 Bath",
      boughtOn: "08/04/2020",
      cost: 379900,
      equity: 20000
    },
    {
      name: "870 S 39th St",
      bedBath: "3 Bed, 1.5 Bath",
      boughtOn: "08/05/2020",
      cost: 375000,
      equity: 23000
    },
    {
      name: "1423 Main St",
      bedBath: "1 Bed, 1 Bath",
      boughtOn: "09/01/2020",
      cost: 750000,
      equity: 34000
    },
    {
      name: "817 Broadway",
      bedBath: "2 Bed, 1 Bath",
      boughtOn: "09/05/2020",
      cost: 820000,
      equity: 60000
    },
    {
      name: "41 N 20th St",
      bedBath: "5 Bed, 3 Bath",
      boughtOn: "09/25/2020",
      cost: 140000,
      equity: 19000
    },
    {
      name: "423 Roman St",
      bedBath: "2 Bed, 2 Bath",
      boughtOn: "11/02/2020",
      cost: 280000,
      equity: 18000
    }
  ],
  showRowHoverBackground: false
};
