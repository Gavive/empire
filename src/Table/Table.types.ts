export type Props<T> = {
  data: T[];
  columns?: TableColumn[];

  /**
   * On hover or selection the background of the row changes as a visual
   * indicator.
   * 
   * @default true
   */
  showRowHoverBackground?: boolean;
};

export type TableColumn = {
  /**
   * Title of the column.
   */
  label: string;

  /**
   * Key to lookup in every data object for the value of each
   * row under the column.
   */
  key: string;
};
