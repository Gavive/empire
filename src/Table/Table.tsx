import React, { Component } from "react";
import styled from "styled-components";
import { Props } from "./Table.types";

const TableContainer = styled.table`
  width: 100%;
  display: table;
  color: ${props => props.theme.color};
  border-spacing: 0;
  border-collapse: collapse;
`;

const TableHead = styled.thead`
  display: table-header-group;
`;

const TableHeadRow = styled.tr`
`;

const TableHeadCell = styled.th`
  font-weight: 600;
  line-height: 1.5rem;
  display: table-cell;
  padding: 16px;
  font-size: 0.875rem;
  text-align: left;
  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
  line-height: 1.43;
  border-bottom: 1px solid rgba(81, 81, 81, 1);
  letter-spacing: 0.01071em;
  vertical-align: inherit;
`;

const TableBody = styled.tbody`
  display: table-row-group;
  vertical-align: middle;
  border-color: inherit;
`;

const TableBodyRow = styled.tr<Partial<Props<any>>>`
  color: inherit;
  display: table-row;
  outline: 0;
  vertical-align: middle;
  :hover {
    background: ${props => {
      if (props.showRowHoverBackground) {
        switch(props.theme && props.theme.name) {
          case "dark":
            return "rgba(255, 255, 255, 0.08)";

          case "light":
            return "rgba(0, 0, 0, 0.08)";
        }
      }

      return "none";
    }}
  }
`;

const TableBodyCell = styled.th`
  display: table-cell;
  padding: 16px;
  font-size: 0.875rem;
  text-align: left;
  font-family: "Roboto", "Helvetica", "Arial", sans-serif;
  font-weight: 400;
  line-height: 1.43;
  border-bottom: 1px solid rgba(81, 81, 81, 1);
  letter-spacing: 0.01071em;
  vertical-align: inherit;
`;

TableBodyRow.defaultProps = {
  showRowHoverBackground: true
};

export default class Table<T> extends Component<Props<T>> {
  constructor(props: Props<T>) {
    super(props);
  }

  private _renderTableHead() {
    return (
      <TableHead>
        <TableHeadRow>
          {
            this.props.columns?.map(col => (
              <TableHeadCell key={col.key}>{col.label}</TableHeadCell>
            ))
          }
        </TableHeadRow>
      </TableHead>
    );
  }

  private _renderTableBody() {
    return (
      <TableBody>
        {
          this.props.data.map((row: T, index: number) => (
            <TableBodyRow key={index} showRowHoverBackground={this.props.showRowHoverBackground}>
              {
                this.props.columns?.map(col => (
                  <TableBodyCell key={`${index}-${col.key}`}>
                    {(row as any)[col.key]}
                  </TableBodyCell>
                ))
              }
            </TableBodyRow>
          ))
        }
      </TableBody>
    );
  }

  public render(): React.ReactNode {
    return (
      <TableContainer>
        {this._renderTableHead()}
        {this._renderTableBody()}
      </TableContainer>
    );
  }
}
