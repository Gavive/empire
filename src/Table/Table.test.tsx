import React from "react";
import { render } from "@testing-library/react";

import { Props } from "./Table.types";
import Table from "./Table";

type TestData = {
  test1: string;
  test2: number;
}

describe("🏓 Table", () => {
  let props: Props<TestData>;

  beforeEach(() => {
    props = {
      data: [
        {
          test1: "TestValue1",
          test2: 0
        },
        {
          test1: "TestValue2",
          test2: 1
        }
      ],
      columns: [
        {
          key: "test1",
          label: "Test1"
        },
        {
          key: "test2",
          label: "Test2"
        }
      ]
    };
  });

  const renderComponent = () => render(<Table {...props} />);

  it("to match snapshot", () => {
    const component = renderComponent();

    expect(component).toMatchSnapshot();
  }) 
});
