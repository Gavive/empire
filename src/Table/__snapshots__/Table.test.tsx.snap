// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`üèì Table to match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        class="sc-AxjAm ctjxld"
      >
        <thead
          class="sc-AxirZ bQauTr"
        >
          <tr
            class="sc-AxiKw cydMkg"
          >
            <th
              class="sc-AxhCb iJvMiA"
            >
              Test1
            </th>
            <th
              class="sc-AxhCb iJvMiA"
            >
              Test2
            </th>
          </tr>
        </thead>
        <tbody
          class="sc-AxhUy bYsaPV"
        >
          <tr
            class="sc-AxgMl gxjrAS"
          >
            <th
              class="sc-AxheI fQlTRg"
            >
              TestValue1
            </th>
            <th
              class="sc-AxheI fQlTRg"
            >
              0
            </th>
          </tr>
          <tr
            class="sc-AxgMl gxjrAS"
          >
            <th
              class="sc-AxheI fQlTRg"
            >
              TestValue2
            </th>
            <th
              class="sc-AxheI fQlTRg"
            >
              1
            </th>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      class="sc-AxjAm ctjxld"
    >
      <thead
        class="sc-AxirZ bQauTr"
      >
        <tr
          class="sc-AxiKw cydMkg"
        >
          <th
            class="sc-AxhCb iJvMiA"
          >
            Test1
          </th>
          <th
            class="sc-AxhCb iJvMiA"
          >
            Test2
          </th>
        </tr>
      </thead>
      <tbody
        class="sc-AxhUy bYsaPV"
      >
        <tr
          class="sc-AxgMl gxjrAS"
        >
          <th
            class="sc-AxheI fQlTRg"
          >
            TestValue1
          </th>
          <th
            class="sc-AxheI fQlTRg"
          >
            0
          </th>
        </tr>
        <tr
          class="sc-AxgMl gxjrAS"
        >
          <th
            class="sc-AxheI fQlTRg"
          >
            TestValue2
          </th>
          <th
            class="sc-AxheI fQlTRg"
          >
            1
          </th>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
